1) Import dependencies

In [35]:
import pandas as pd 
import numpy as np 

2) import all data

Here are some questions about the dataset. I imported the CSV file with pandas 2.0 using this function: df = pd.read_csv('Sport car price.csv', sep=',', Engine='pyarrow', use_nullable_dtypes=True. However, when I examine the dataset, the columns that I need which should have been recognized as integer or float types, are still being recognized as objects, why?

In [36]:
#df = pd.read_csv('Sport car price.csv', sep=',',engine='pyarrow',use_nullable_dtypes=True)

In [37]:
#df.info()

In [38]:
df = pd.read_csv('Sport car price.csv', sep=',')

In [39]:
df.head()

Car Make Car Model  Year Engine Size (L) Horsepower Torque (lb-ft)   
0      Porsche       911  2022               3        379            331  \
1  Lamborghini   Huracan  2021             5.2        630            443   
2      Ferrari   488 GTB  2022             3.9        661            561   
3         Audi        R8  2022             5.2        562            406   
4      McLaren      720S  2021               4        710            568   

  0-60 MPH Time (seconds) Price (in USD)  
0                       4        101,200  
1                     2.8        274,390  
2                       3        333,750  
3                     3.2        142,700  
4                     2.7        298,000

In [40]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1007 entries, 0 to 1006
Data columns (total 8 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Car Make                 1007 non-null   object
 1   Car Model                1007 non-null   object
 2   Year                     1007 non-null   int64 
 3   Engine Size (L)          997 non-null    object
 4   Horsepower               1007 non-null   object
 5   Torque (lb-ft)           1004 non-null   object
 6   0-60 MPH Time (seconds)  1007 non-null   object
 7   Price (in USD)           1007 non-null   object
dtypes: int64(1), object(7)
memory usage: 63.1+ KB


3) analyze all data, how many null spaces I have, the type of our colums and change someting if I need do it.
- check why I can't change the type of "horsepower" "torque" "0-60 MPH" and "price"  

In [41]:
df['Horsepower'].unique()

array(['379', '630', '661', '562', '710', '617', '523', '490', '760',
       '600', '1500', '717', '296', '1280', '471', '416', '454', '300',
       '505', '320', '626', '671', '622', '720', '1914', '414', '759',
       '986', '591', '503', '650', '660', '350', '641', '611', '394',
       '612', '369', '603', '455', '460', '325', '349', '592', '444',
       '405', '797', '770', '332', '473', '480', '573', '380', '1600',
       '181', '620', '764', '624', '1000+', '382', '800', '715', '690',
       '730', '469', '365', '401', '645', '435', '1020', '500', '780',
       '750', '402', '575', '729', '789', '577', '495', '237', '310',
       '791', '1874', '542', '368', '616', '1479', '755', '1,000+', '288',
       '1888', '10000+', '482', '1973', '1262', '1035', '819', '385',
       '647', '1200', '1578', '625', '583', '429', '563', '400', '707',
       '887', '1972', '305', '640', '255', '689', '372', '1000', '2000',
       '550', '10,000', '1,500', '10,000+', '485', '1,020', '1872', '621'

In [42]:
df['Torque (lb-ft)'].unique()

array(['331', '443', '561', '406', '568', '553', '494', '465', '625',
       '481', '516', '1180', '656', '295', '1015', '398', '317', '384',
       '280', '243', '664', '531', '468', '737', '738', '1696', '309',
       '590', '479', '650', '550', '276', '626', '369', '420', '627',
       '455', '505', '560', '457', '707', '270', '354', '476', '339',
       '1106', '151', '605', '368', '723', '642', '509', '604', '507',
       '513', '600', '440', '1050', '708', '774', '254', '663', '332',
       '530', '470', '258', '290', '413', '1732', '376', '-', '10,000+',
       '236', '0', '472', '1254', '848', '1300', '442', '641', '498',
       '350', nan, '944', '268', '184', '400', '263', '7,376', '1,180',
       '475', '1,050', '740', '538'], dtype=object)

In [43]:
df['0-60 MPH Time (seconds)'].unique()

array(['4', '2.8', '3', '3.2', '2.7', '3.1', '3.8', '3.5', '2.5', '2.4',
       '5.4', '4.4', '4.8', '4.7', '3.6', '4.1', '1.85', '4.5', '3.3',
       '3.9', '4.2', '3.4', '5.1', '4.3', '2.9', '5', '5.3', '4.9', '6.5',
       '3.7', '1.95', '1.9', '1.98', '2.6', '2.3', '4.6', '< 1.9', '1.8',
       '2.1', '5.2', '2.2', '6.4', '2'], dtype=object)

In [44]:
df['Price (in USD)'] = df['Price (in USD)'].str.replace(',','')
df['Price (in USD)'] = df['Price (in USD)'].astype('int')

now I understand:
- horsepower have fake data, I'm going to analyse the column. 


I'm going to view all data that have as a fake data and edit that information, on this dataset i would have in horsepower values in range of 300 and 1479 and i have:
- 1500, 1914, 1600, 1888, 1973, 1578, 1972, 2000 and 1872
I have too numbers like these:
- 1,020, 1,500 and 10,000
And another data like:
- 1000+, 1,000+, 10000+ and 10,000+
that data don't let me work, so I have 3 options, or delete the columns/rows that have that data, edit it or try to understand my dataset and do something diferent but trying to not delete my data, like analyse that information.

In [45]:
df.groupby('Horsepower').get_group('10,000+')

Car Make Car Model  Year Engine Size (L) Horsepower Torque (lb-ft)   
916    Tesla  Roadster  2022             NaN    10,000+            NaN  \

    0-60 MPH Time (seconds)  Price (in USD)  
916                     1.9          200000

In [46]:
df.groupby('Horsepower').get_group('1,000+')

Car Make Car Model  Year Engine Size (L) Horsepower Torque (lb-ft)   
364    Tesla  Roadster  2023        Electric     1,000+            737  \

    0-60 MPH Time (seconds)  Price (in USD)  
364                   < 1.9          200000

In [47]:
# It's weird that I have 2 fake data from tesla.
tesla = df.groupby('Car Make').get_group('Tesla')

On tesla I have a lot fake and repeated data, if I clean that I'll earn time.

In [48]:
# but I have to view all that data or maybe I'll miss something
df.groupby('Horsepower').get_group('10000+')

Car Make Car Model  Year Engine Size (L) Horsepower Torque (lb-ft)   
389    Tesla  Roadster  2022             NaN     10000+              0  \

    0-60 MPH Time (seconds)  Price (in USD)  
389                     1.9          200000

In [49]:
df.groupby('Horsepower').get_group('1000+')

Car Make Car Model  Year Engine Size (L) Horsepower Torque (lb-ft)   
99     Tesla  Roadster  2022        Electric      1000+            737  \
335    Tesla  Roadster  2022               -      1000+              -   
354    Tesla  Roadster  2022        Electric      1000+        10,000+   

    0-60 MPH Time (seconds)  Price (in USD)  
99                      1.9          200000  
335                     1.9          200000  
354                     1.9          200000

all the data that have x number with "+" its from tesla, so, I'll delete rows and fill the others spaces.

In [50]:
tesla.sort_values('Year')

Car Make      Car Model  Year       Engine Size (L) Horsepower   
401    Tesla  Model S Plaid  2021  Electric (tri-motor)       1020  \
171    Tesla  Model S Plaid  2021                   NaN       1020   
320    Tesla  Model S Plaid  2021              Electric       1020   
642    Tesla  Model S Plaid  2021              Electric       1020   
511    Tesla  Model S Plaid  2021              Electric       1020   
885    Tesla       Roadster  2022                     0     10,000   
821    Tesla       Roadster  2022              Electric       1000   
622    Tesla  Model S Plaid  2022    Electric (100 kWh)       1020   
540    Tesla  Model S Plaid  2022              Electric       1020   
527    Tesla  Model S Plaid  2022              Electric       1020   
99     Tesla       Roadster  2022              Electric      1000+   
389    Tesla       Roadster  2022                   NaN     10000+   
354    Tesla       Roadster  2022              Electric      1000+   
335    Tesla       Roadster  2022                     -      1000+   
300    Tesla        Model S  2022              Electric       1020   
247    Tesla  Model S Plaid  2022                   NaN       1020   
916    Tesla       Roadster  2022                   NaN    10,000+   
965    Tesla  Model S Plaid  2022              Electric      1,020   
364    Tesla       Roadster  2023              Electric     1,000+   

    Torque (lb-ft) 0-60 MPH Time (seconds)  Price (in USD)  
401           1050                     1.9          141190  
171           1050                    1.98          131190  
320           1050                     1.9          131100  
642            NaN                     1.9          139990  
511           1050                     1.9          141990  
885          7,376                     1.9          200000  
821            737                     1.9          200000  
622           1050                    1.98          131190  
540           1050                    1.98          139990  
527           1050                    1.98          140000  
99             737                     1.9          200000  
389              0                     1.9          200000  
354        10,000+                     1.9          200000  
335              -                     1.9          200000  
300           1050                    1.98          119000  
247           1050                     1.9          131190  
916            NaN                     1.9          200000  
965          1,050                     1.9          141190  
364            737                   < 1.9          200000

In [51]:
index = [99, 171,247,320, 335, 354, 389, 401, 540, 622, 642, 885, 916,965]

In [52]:
df.drop(index, inplace= True)

In [53]:
df.loc[df['Horsepower']=='1,000+','Horsepower'] = '1341'

In [54]:
df.loc[df['0-60 MPH Time (seconds)']=='< 1.9', '0-60 MPH Time (seconds)' ] = 1.9

In [55]:
df.groupby('Car Make').get_group('Tesla')

Car Make      Car Model  Year Engine Size (L) Horsepower Torque (lb-ft)   
300    Tesla        Model S  2022        Electric       1020           1050  \
364    Tesla       Roadster  2023        Electric       1341            737   
511    Tesla  Model S Plaid  2021        Electric       1020           1050   
527    Tesla  Model S Plaid  2022        Electric       1020           1050   
821    Tesla       Roadster  2022        Electric       1000            737   

    0-60 MPH Time (seconds)  Price (in USD)  
300                    1.98          119000  
364                     1.9          200000  
511                     1.9          141990  
527                    1.98          140000  
821                     1.9          200000

now I'm going to analyse the another info

In [56]:
df.groupby('Horsepower').get_group('1500')

Car Make         Car Model  Year Engine Size (L) Horsepower   
11   Bugatti            Chiron  2021               8       1500  \
85   Bugatti            Chiron  2022               8       1500   
113  Bugatti            Chiron  2021               8       1500   
158  Bugatti            Chiron  2021               8       1500   
206  Bugatti            Chiron  2021               8       1500   
274  Bugatti            Chiron  2021               8       1500   
341  Bugatti            Chiron  2021               8       1500   
376  Bugatti            Chiron  2022               8       1500   
434  Bugatti            Chiron  2022               8       1500   
499  Bugatti            Chiron  2022               8       1500   
519  Bugatti            Chiron  2021               8       1500   
624  Bugatti  Chiron Pur Sport  2021               8       1500   
631  Bugatti            Chiron  2021               8       1500   
683  Bugatti            Chiron  2022               8       1500   
983  Bugatti            Chiron  2022               8       1500   

    Torque (lb-ft) 0-60 MPH Time (seconds)  Price (in USD)  
11            1180                     2.4         3000000  
85            1180                     2.4         3000000  
113           1180                     2.4         3000000  
158           1180                     2.4         3000000  
206           1180                     2.3         3000000  
274           1180                     2.4         2998000  
341           1180                     2.4         3000000  
376           1180                     2.4         3000000  
434           1180                     2.4         3000000  
499           1180                     2.3         3000000  
519           1180                     2.3         3000000  
624           1180                     2.3         3599000  
631           1180                     2.5         3000000  
683           1180                     2.4         3000000  
983           1180                     2.5         3900000

I have 2 problems now, first the range of Horsepower isn't from 300 to 1479 second when i saw that, the dataset will probably has a lot repeated data, and that it's a problem, I don't have an idea how many data repeated i have. 

In [57]:
df.groupby('Horsepower').get_group('1914')

Car Make Car Model  Year Engine Size (L) Horsepower Torque (lb-ft)   
26     Rimac    Nevera  2022        Electric       1914           1696  \
97     Rimac    Nevera  2022        Electric       1914           1696   
168    Rimac     C_Two  2022             NaN       1914           1696   
278    Rimac     C_Two  2022        Electric       1914           1732   
352    Rimac    Nevera  2022        Electric       1914           1696   
439    Rimac     C_Two  2021        Electric       1914           1732   
509    Rimac     C_Two  2021        Electric       1914           1696   
526    Rimac     C_Two  2022        Electric       1914           1696   
640    Rimac    Nevera  2021        Electric       1914           1696   
686    Rimac     C_Two  2022             NaN       1914           1696   
824    Rimac    Nevera  2021        Electric       1914           1696   
986    Rimac    Nevera  2022        Electric       1914           1696   

    0-60 MPH Time (seconds)  Price (in USD)  
26                     1.85         2400000  
97                     1.95         2400000  
168                     1.9         2400000  
278                    1.85         2400000  
352                    1.85         2400000  
439                     1.8         2400000  
509                     1.9         2400000  
526                     1.9         2400000  
640                     1.9         2400000  
686                    1.85         2400000  
824                    1.85         2400000  
986                    1.85         2400000

now I don't know witch data it's real or not, and I can't check all the information, but if i search the car make and their torque, the data is right, so I'll delete the nan types from my dataset, I can do it in this case because I have a lot of repeated data and at least one of them have the complete information

In [58]:
df.dropna()

Car Make Car Model  Year Engine Size (L) Horsepower Torque (lb-ft)   
0         Porsche       911  2022               3        379            331  \
1     Lamborghini   Huracan  2021             5.2        630            443   
2         Ferrari   488 GTB  2022             3.9        661            561   
3            Audi        R8  2022             5.2        562            406   
4         McLaren      720S  2021               4        710            568   
...           ...       ...   ...             ...        ...            ...   
1002   Koenigsegg     Jesko  2022               5       1280           1106   
1003        Lotus     Evija  2021  Electric Motor       1972           1254   
1004      McLaren     Senna  2021               4        789            590   
1005       Pagani    Huayra  2021               6        764            738   
1006        Rimac    Nevera  2021  Electric Motor       1888           1696   

     0-60 MPH Time (seconds)  Price (in USD)  
0                          4          101200  
1                        2.8          274390  
2                          3          333750  
3                        3.2          142700  
4                        2.7          298000  
...                      ...             ...  
1002                     2.5         3000000  
1003                       2         2000000  
1004                     2.7         1000000  
1005                       3         2600000  
1006                    1.85         2400000  

[986 rows x 8 columns]

In [59]:
df['Horsepower'] = df['Horsepower'].str.replace(',','')

In [60]:
df['Horsepower'] = df['Horsepower'].astype('int')

In [61]:
df['Torque (lb-ft)'].replace('N/A', np.nan, inplace=True)

In [62]:
df['Torque (lb-ft)'] = df['Torque (lb-ft)'].str.replace(',','')

In [63]:
df['Torque (lb-ft)'] = df['Torque (lb-ft)'].astype('float')

In [64]:
df['0-60 MPH Time (seconds)'] = df['0-60 MPH Time (seconds)'].astype('float')

In [65]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 993 entries, 0 to 1006
Data columns (total 8 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Car Make                 993 non-null    object 
 1   Car Model                993 non-null    object 
 2   Year                     993 non-null    int64  
 3   Engine Size (L)          987 non-null    object 
 4   Horsepower               993 non-null    int32  
 5   Torque (lb-ft)           992 non-null    float64
 6   0-60 MPH Time (seconds)  993 non-null    float64
 7   Price (in USD)           993 non-null    int32  
dtypes: float64(2), int32(2), int64(1), object(3)
memory usage: 62.1+ KB


now I'm going to delete all the data duplicated, and I'll have a dataset with 100% of the information that I need.

In [66]:
df.drop_duplicates(subset=['Car Make', 'Car Model', 'Year'],inplace=True)

In [67]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 262 entries, 0 to 998
Data columns (total 8 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Car Make                 262 non-null    object 
 1   Car Model                262 non-null    object 
 2   Year                     262 non-null    int64  
 3   Engine Size (L)          260 non-null    object 
 4   Horsepower               262 non-null    int32  
 5   Torque (lb-ft)           262 non-null    float64
 6   0-60 MPH Time (seconds)  262 non-null    float64
 7   Price (in USD)           262 non-null    int32  
dtypes: float64(2), int32(2), int64(1), object(3)
memory usage: 16.4+ KB


Now I have my data cleaned, I'm going to analyse the data on power bi and try to find some information about the dataset.

In [68]:
df.to_csv('SportCars.csv', index=False)